src/utils.py

```diff
@@ -47,91 +72,112 @@ def read_path_if_provided(spark, path):
     if parquet_files:
         return spark.read.parquet(path)
 
-def add_delta(
-        df: pd.DataFrame,
-        metric: str,
-        previous_run: str,
-        latest_run: str
-    ):
-
-    df[f"Δ in number of {metric}"] = (
-        df[f"Nr of {metric} in {latest_run.split('-')[0]}"]
-        .sub(
-            df[f"Nr of {metric} in {previous_run.split('-')[0]}"],
-            axis=0, fill_value=0))
-
-def compare_entity(
-    df: pd.DataFrame,
-    entity_name: str,
-    latest_run: str,
-    previous_run: str
-) -> pd.DataFrame:
-
-    if entity_name in ['diseases', 'drugs', 'targets']:
+
+def add_delta(df: pd.DataFrame, metric: str, previous_run: str, latest_run: str):
+    """
+    It takes a dataframe, a metric, and two run names, and adds a column to the dataframe that shows the
+    difference in the number of that metric between the two runs
+
+    Args:
+      df (pd.DataFrame): the dataframe to add the delta column to
+      metric (str): the metric we're interested in
+      previous_run (str): the name of the previous run
+      latest_run (str): the latest run of the pipeline
+    """
+
+    df[f"Δ in number of {metric}"] = df[f"Nr of {metric} in {latest_run.split('-')[0]}"].sub(
+        df[f"Nr of {metric} in {previous_run.split('-')[0]}"], axis=0, fill_value=0
+    )
+
+
+def compare_entity(df: pd.DataFrame, entity_name: str, latest_run: str, previous_run: str) -> pd.DataFrame:
+    """
+    It takes a dataframe with metrics and returns a dataframe with the difference between the latest and previous run.
+
+    Args:
+      df (pd.DataFrame): the dataframe to be modified
+      entity_name (str): the name of the entity you want to compare.
+      latest_run (str): the name of the latest run
+      previous_run (str): the name of the previous run
+
+    Returns:
+      A dataframe with the number of entities in the latest run and the difference between the latest
+    and previous run.
+    """
+    if entity_name in {'diseases', 'drugs', 'targets'}:
         add_delta(df, entity_name, previous_run, latest_run)
+    if entity_name == 'associations':
+        add_delta(df, "direct associations", previous_run, latest_run)
```

Comment:
I don't understand what is happening here. The `add_delta` has no return value. Does the `add_delta` in place modifies df? How the df at the end of the function changed?